controlplane/src/core/services/SchemaGraphPruner.ts

```diff
@@ -0,0 +1,348 @@
+  }): Promise<GraphPruningIssueResult[]> => {
+    const allFields = this.getAllFields({});
+    const fieldsInGracePeriod = await this.subgraphRepo.getSubgraphFieldsInGracePeriod({ subgraphId, namespaceId });
+
+    const fieldsToBeChecked = allFields.filter((field) => {
+      return !fieldsInGracePeriod.some((f) => f.path === field.path) && !addedFields.some((f) => f.path === field.path);
+    });
+
+    const graphPruningIssues: GraphPruningIssueResult[] = [];
+
+    for (const federatedGraph of federatedGraphs) {
+      const unusedFieldsWithTypeNames = await this.usageRepo.getUnusedFields({
+        fields: fieldsToBeChecked,
+        organizationId,
+        federatedGraphId: federatedGraph.id,
+        range: rangeInDays * 24,
+      });
+      const unusedFields = allFields.filter((field) =>
+        unusedFieldsWithTypeNames.some((f) => f.name === field.name && f.typeName === field.typeName),
+      );
+      for (const field of unusedFields) {
+        graphPruningIssues.push({
+          graphPruningRuleType: 'UNUSED_FIELDS',
+          severity: severityLevel === 'error' ? LintSeverity.error : LintSeverity.warn,
+          fieldPath: field.path,
+          message: `Field ${field.name} of type ${field.typeName} has not used in the past ${rangeInDays} days`,
+          issueLocation: {
+            line: field.location.line || 0,
+            column: field.location.column || 0,
+            endLine: field.location.endLine,
+            endColumn: field.location.endColumn,
+          },
+          federatedGraphId: federatedGraph.id,
+          federatedGraphName: federatedGraph.name,
+        });
+      }
+    }
+
+    return graphPruningIssues;
+  };
+
+  fetchDeprecatedFields = async ({
+    subgraphId,
+    namespaceId,
+    organizationId,
+    federatedGraphs,
+    rangeInDays,
+    severityLevel,
+    addedDeprecatedFields,
+  }: {
+    federatedGraphs: FederatedGraphDTO[];
+    subgraphId: string;
+    namespaceId: string;
+    organizationId: string;
+    rangeInDays: number;
+    severityLevel: LintSeverityLevel;
+    addedDeprecatedFields: SchemaDiff[];
+  }): Promise<GraphPruningIssueResult[]> => {
+    const allDeprecatedFields = this.getAllFields({ onlyDeprecated: true });
+    const deprecatedFieldsInGracePeriod = await this.subgraphRepo.getSubgraphFieldsInGracePeriod({
+      subgraphId,
+      namespaceId,
+      onlyDeprecated: true,
+    });
+
+    const deprecatedFieldsToBeChecked = allDeprecatedFields.filter((field) => {
+      return (
+        !deprecatedFieldsInGracePeriod.some((f) => f.path === field.path) &&
+        !addedDeprecatedFields.some((f) => f.path === field.path)
+      );
+    });
+
+    if (deprecatedFieldsToBeChecked.length === 0) {
+      return [];
+    }
+
+    const graphPruningIssues: GraphPruningIssueResult[] = [];
+
+    for (const federatedGraph of federatedGraphs) {
+      const usedDeprecatedFieldsWithTypeNames = await this.usageRepo.getUsedFields({
```

Comment:
I'd run these in parallel. You can use a library like [p-limit](https://www.npmjs.com/package/p-limit) to limit max concurrency.

Reply:
Done